airbyte-webapp/src/hooks/services/Connection/ConnectionFormService.tsx

```diff
@@ -0,0 +1,117 @@
+import { createFormErrorMessage } from "utils/errorStatusMessage";
+import { ConnectionFormMode, ConnectionFormSubmitResult } from "views/Connection/ConnectionForm/ConnectionForm";
+import {
+  ConnectionFormValues,
+  connectionValidationSchema,
+  FormikConnectionFormValues,
+  mapFormPropsToOperation,
+  useFrequencyDropdownData,
+  useInitialValues,
+} from "views/Connection/ConnectionForm/formConfig";
+
+import { useFormChangeTrackerService, useUniqueFormId } from "../FormChangeTracker";
+import { ModalCancel } from "../Modal";
+
+export interface ConnectionServiceProps {
+  connection:
+    | WebBackendConnectionRead
+    | (Partial<WebBackendConnectionRead> & Pick<WebBackendConnectionRead, "syncCatalog" | "source" | "destination">);
+  mode: ConnectionFormMode;
+  onSubmit: (values: ConnectionFormValues) => Promise<ConnectionFormSubmitResult | void>;
+  onAfterSubmit?: () => void;
+  onFrequencySelect?: (item: DropDownRow.IDataItem) => void;
+  onCancel?: () => void;
+}
+
+const useConnectionForm = ({
+  connection,
+  mode,
+  onSubmit,
+  onAfterSubmit,
+  onFrequencySelect,
+  onCancel,
+}: ConnectionServiceProps) => {
+  // Need to track form dirty state
+  // Might be easy if form context is in here
+  const [submitError, setSubmitError] = useState<Error | null>(null);
+  const workspaceId = useCurrentWorkspaceId();
+  const { clearFormChange } = useFormChangeTrackerService();
+  const formId = useUniqueFormId();
+
+  const destDefinition = useGetDestinationDefinitionSpecification(connection.destination.destinationDefinitionId);
+  const initialValues = useInitialValues(connection, destDefinition, mode !== "create");
+
+  const onFormSubmit = useCallback(
+    async (values: FormikConnectionFormValues, formikHelpers: FormikHelpers<FormikConnectionFormValues>) => {
+      // Set the scheduleType based on the schedule value
+      values["scheduleType"] = values.scheduleData?.basicSchedule
+        ? ConnectionScheduleType.basic
+        : ConnectionScheduleType.manual;
+
+      const formValues: ConnectionFormValues = connectionValidationSchema.cast(values, {
+        context: { isRequest: true },
+      }) as unknown as ConnectionFormValues; // TODO: We should align these types
+
+      formValues.operations = mapFormPropsToOperation(values, connection.operations, workspaceId);
+
+      setSubmitError(null);
+      try {
+        await onSubmit(formValues);
+
+        formikHelpers.resetForm({ values });
+        clearFormChange(formId);
+
+        onAfterSubmit?.();
+      } catch (e) {
+        if (!(e instanceof ModalCancel)) {
+          setSubmitError(e);
+        }
+      }
+    },
+    [connection.operations, workspaceId, onSubmit, clearFormChange, formId, onAfterSubmit]
+  );
+
+  const errorMessage = useMemo(() => (submitError ? createFormErrorMessage(submitError) : null), [submitError]);
+  const frequencies = useFrequencyDropdownData(connection.scheduleData);
+
+  const formDirty = new Subject<boolean>();
```

Comment:
I think it's odd to expose the observable in the hook. Preferably the hook should provide some interface that simplifies the flow and internally use the observable to track.

Reply:
Removed!